flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/reader/CassandraSplitReader.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsChange;
+import org.apache.flink.connector.cassandra.source.split.CassandraSplit;
+import org.apache.flink.connector.cassandra.source.split.CassandraSplitState;
+import org.apache.flink.connector.cassandra.source.split.RingRange;
+import org.apache.flink.streaming.connectors.cassandra.ClusterBuilder;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.ColumnMetadata;
+import com.datastax.driver.core.Metadata;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.Token;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * {@link SplitReader} for Cassandra source. This class is responsible for fetching the records as
+ * {@link CassandraRow}. For that, it executes a range query (query that outputs records belonging
+ * to a {@link RingRange}) based on the user specified query. This class manages the Cassandra
+ * cluster and session.
+ */
+public class CassandraSplitReader implements SplitReader<CassandraRow, CassandraSplit> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CassandraSplitReader.class);
+    public static final String SELECT_REGEXP = "(?i)select .+ from (\\w+)\\.(\\w+).*;$";
+
+    private final Cluster cluster;
+    private final Session session;
+    private final Set<CassandraSplitState> unprocessedSplits;
+    private final AtomicBoolean wakeup = new AtomicBoolean(false);
+    private final String query;
+
+    public CassandraSplitReader(ClusterBuilder clusterBuilder, String query) {
+        // need a thread safe set
+        this.unprocessedSplits = ConcurrentHashMap.newKeySet();
+        this.query = query;
+        cluster = clusterBuilder.getCluster();
+        session = cluster.connect();
+    }
+
+    @Override
+    public RecordsWithSplitIds<CassandraRow> fetch() {
+        Map<String, Collection<CassandraRow>> recordsBySplit = new HashMap<>();
+        Set<String> finishedSplits = new HashSet<>();
+        Metadata clusterMetadata = cluster.getMetadata();
+
+        String partitionKey = getPartitionKey(clusterMetadata);
+        String finalQuery = generateRangeQuery(query, partitionKey);
+        PreparedStatement preparedStatement = session.prepare(finalQuery);
+        // Set wakeup to false to start consuming.
+        wakeup.compareAndSet(true, false);
+        for (CassandraSplitState cassandraSplitState : unprocessedSplits) {
```

Comment:
No. `CassandraSplits` are assigned to `CassandraSplitReader` (see `CassandraSplitEnumerator#assignUnprocessedSplitsToReader`). A `CassandraSplitReader` might be assigned several splits in case the **strange** splitGenerator returns more splits than requested parallelism. With the new `splitGenerator` there will be exactly as many splits as the parallelism.  So here we read all the records that belong to **the splits assigned to the reader***.

Reply:
Of course the splits are scoped to the reader.
But say, if parallelism=1, then you're reading all records from all splits into memory.
Even with a higher parallelism cases I don't see this working very well for larger tables.

While the split reader API docs are a bit unclear on this, you _are_ allowed to exist the fetch call early. If you adjust the splits accordingly fetch() will be called again later on and you can continue where you left off.

Reply:
> from all splits into memory

Well, in 99% of the cases, `unprocessedSplits` will contain only a single split as the enumerator now assigns only one split per reader. The remaining case I see is when the parallelism of the job is decreased (either by user or failover). In that case, as you know, upon restoration, a reader could be assigned the splits of another reader (as there are less readers after parallelism decrease).

But still you're right, in case of parallelism=1 then, all the records would be in 1 split in 1 reader, there will be only a single query issued and a single resultset containing all the records. These records will indeed be stored in `recordsBySplit` to serve for constructing returned `RecordsBySplits`. That being said, as there will be a single resultset, we need to figure out a way of splitting it for that particular case. Luckily Cassandra `ResultSet` is paginated with results fetched only when all the records of the previous page were consumed. So, I could store in the `CassandraSplitState` a reference to the `ResultSet` to resume the output of remaining records on a later `fetch()`. My question is: what would be the condition to exit fetch() earlier? What about allowing the user to configure a `maxRecordPerSplit` ?

It is true that fetch early exit should be mentioned in the docs. I'll open a ticket for that and do the PR later.

Reply:
https://issues.apache.org/jira/browse/FLINK-30802. Can you assign it to me ?

Reply:
I have pushed an impl that keeps track of the state of the fetch with CassandraSplitState#resultSet. It supports fetch() of MAX_RECORDS_PER_SPLIT and fetch() resume. I added MAX_RECORDS_PER_SPLIT conf parameter as proposed above because I did not know if such parameter existed already. Feel free to point me to an existing one or to propose another parameter name.

Reply:
> I added MAX_RECORDS_PER_SPLIT conf parameter as proposed above because I did not know if such parameter existed already. Feel

I'll look around. The naming seems off though; it's not a maximum records for a split (which would be more relevant for the `SplitGenerator`), but more of a `batch size`.

> So, I could store in the CassandraSplitState a reference to the ResultSet to resume the output of remaining records on a later fetch().

You also have to consider how this information can be stored in a checkpoint. A checkpoint may happen between 2 fetches for 1 split (== first half of emitted will be remembered by Flink), and then the job is restarted. As-is you'd read the first half again, because your split state isn't written to checkpoints.

Not sure how to solve that. If cassandra has a deterministic order in which it returns records (it probably doesnt :see_no_evil: ) then you could remember the count and skip so many records or adjust the start token of the split.

Alternatively we could consider handling this in the SplitGenerator, generating as many splits as necessary to keep at most N records in memory in each split reader. (We'd never have the case that 1 splits contains the entire table)

Reply:
> I'll look around. The naming seems off though; it's not a maximum records for a split (which would be more relevant for the SplitGenerator), but more of a batch size.

Agree, it is a batch size with the current code

> You also have to consider how this information can be stored in a checkpoint

Yes, I did not have time to address this point before my vacation but I wanted to validate with you the general approach of these 3 big architecture changes. Now regarding this point: indeed, the ResultSet is just a handler on the fetching process, so storing it as part of the checkpoint does not help in knowing what was already processed after restoration. That is sure that Cassandra does not provide deterministic order because ordering does not scale, so counting does not work either. I looked into `ResultSet#ExecutionInfo` and found no counter or processed Row ids etc... 

So I guess that leaves the SplitGenerator solution you proposed. In that case the conf becomes a max_rows_per_split and not a batch_size anymore. Also, with this solution: 

- We might get more splits than the current parallelism. So each split reader will read more than one split

- The aim here is to avoid storing too much in memory so I'd prefer that the user conf parameter is in bytes. and I can get my size estimates from Beam 2017 code so that the SplitGenerator provides splits of a given size. Then all the records of each split will be stored in memory to create the needed `RecordsBySplits`. 

WDYT ?

Reply:
> I'd prefer that the user conf parameter is in bytes.

I agree. (well it should be a `MemorySize`, but the same idea)

> I can get my size estimates from Beam 2017 code so that the SplitGenerator provides splits of a given size. Then all the records of each split will be stored in memory to create the needed RecordsBySplits.

Sounds good. We do something similar in the [mongodb connector](https://github.com/apache/flink-connector-mongodb/blob/main/flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSampleSplitter.java).

Reply:
> it should be a MemorySize

Yes that is what I meant.

> Sounds good. We do something similar in the [mongodb connector](https://github.com/apache/flink-connector-mongodb/blob/main/flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSampleSplitter.java).

Yes split from Beam 2017 did something similar: it generates CassandraSplits based on Cassandra cluster partitioner and cluster statistics. It estimates the total size of the table using Cassandra system table system.size_estimates and generates estimateTableSize() / maxSplitMemorySize splits with token boundaries. But there is no way to estimate the size of the data with the optional SQL filters without reading the data. So the splits can be smaller than maxSplitMemorySize.